### ISSUE TITLE
clang: -Wuninitialized false positive when accessing a virtual base class member in a member initializer of an abstract class constructor

### ISSUE THREAD

**Issue by steffen-kiess** [llvm/llvm-project#204858]
The code

```c++
struct C1 {
  C1() {}

  int foo() { return 0; }
};

struct C2 : virtual C1 {
  int i;

  C2() : i(this->foo()) {}

  virtual void makeAbstract() = 0;
};

struct C3 : C2 {
  void makeAbstract() {}
};

int main() { C3 c; }
```

will cause the warning

```
file.cpp:10:18: warning: base class 'C1' is uninitialized when used here to access 'C1::foo' [-Wuninitialized]
   10 |   C2() : i(this->foo()) {}
      |                  ^
```

even though C1, being a base class of C2, should always be initialized before the non-static data members of C2 are initialized (according to my understanding of [class.base.init] in the C++11 standard).

Note that making C2 non-abstract will remove the warning, even though it should not change anything.

Also, changing the C2 constructor to `C2() : C1(), i(this->foo()) {}` will remove the warning, even though the call to `C1()` will always be ignored because C2 is abstract.

**Comment by steffen-kiess** [comment]
Note: I also submitted a similar (but slightly different) bug report for gcc at <https://gcc.gnu.org/bugzilla/show_bug.cgi?id=125906>.
